src/genomicsdb.pyx

```diff
@@ -305,6 +312,78 @@ cdef class _GenomicsDB:
                                              as_ranges(row_ranges))
       
       return pandas.DataFrame(processor.construct_data_frame()).replace(np.nan, '').replace(-99999, '');
+
+    def query_variant_calls_arrow(self,
+                                  array=None,
+                                  column_ranges=None,
+                                  row_ranges=None,
+                                  query_protobuf: query_pb.QueryConfiguration=None,
+                                  non_blocking=False): 
+      """ Query for variant calls from the GenomicsDB workspace using array, column_ranges and row_ranges for subsetting """
+
+      cdef ArrowVariantCallProcessor processor
+      cdef void *arrow_array = NULL
+      cdef void *arrow_schema = NULL
+
+      if non_blocking:
+        processor.set_threaded(1)
+
+      def query_calls():
+        if query_protobuf:
+          if array or column_ranges or row_ranges:
+            raise GenomicsDBException("Cannot specify query_protobuf and array/column_ranges/row_ranges together")
+          configstring = as_protobuf_string(query_protobuf.SerializeToString())
+          with nogil:
+            self._genomicsdb.query_variant_calls(processor, configstring, GENOMICSDB_PROTOBUF_BINARY_STRING)
+        elif array is None:
+          configstring = as_string("")
+          with nogil:
+            self._genomicsdb.query_variant_calls(processor, configstring, GENOMICSDB_NONE)
+        elif column_ranges is None:
+          configstring = as_string(array)
+          rows = scan_full()
+          with nogil:
+            self._genomicsdb.query_variant_calls(processor, configstring, rows)
+        elif row_ranges is None:
+          configstring = as_string(array)
+          columns = as_ranges(column_ranges)
+          with nogil:
+            self._genomicsdb.query_variant_calls(processor, configstring, columns)
+        else:
+          configstring = as_string(array)
+          columns = as_ranges(column_ranges)
+          rows = as_ranges(row_ranges)
+          with nogil:
+            self._genomicsdb.query_variant_calls(processor, configstring,
+                                                 columns, rows)
+
+      if non_blocking:
+        query_thread = threading.Thread(target=query_calls)
+        query_thread.start()
+      else:
+        query_calls()
+
+      cdef object schema_capsule
+      while True:
+        arrow_array = processor.arrow_array()
+        if arrow_schema == NULL:
```

Comment:
Is this first block only to initialize and yield the schema on the first iteration? If so, could it be moved out of the `while` block?

Reply:
We can get to the schema only when arrow_array() returns the first time. I could add another separate set of semaphores for arrow_schema to wait on if necessary. See `genomicsdb_nanoarrow_processor.cc` in https://github.com/GenomicsDB/GenomicsDB/pull/338.